powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/encryptor/ecies/ServerEciesEncryptor.java

```diff
@@ -0,0 +1,265 @@
+            envelopeKey = new EciesEnvelopeKey(encryptorSecrets.getEnvelopeKey(), ephemeralPublicKey);
+        } else {
+            // Derive shared secret from private key and ephemeral public key.
+            envelopeKey = EciesEnvelopeKey.fromPrivateKey(
+                    encryptorSecrets.getServerPrivateKey(),
+                    ephemeralPublicKey,
+                    encryptorId.getEciesSharedInfo1(encryptorParameters.getProtocolVersion())
+            );
+        }
+        // Return secrets object with the precalculated keys.
+        return new ServerEncryptorSecrets(envelopeKey.getSecretKey(), sharedInfo2Base);
+    }
+
+    @Override
+    public byte[] decryptRequest(EncryptedRequest request) throws EncryptorException {
+        if (!canDecryptRequest()) {
+            throw new EncryptorException("Encryptor is not ready for request decryption.");
+        }
+        // Validate and decode request payload
+        if (!validator.validateEncryptedRequest(request)) {
+            throw new EncryptorException("Invalid encrypted request object.");
+        }
+
+        final byte[] ephemeralPublicKey = Base64.getDecoder().decode(request.getEphemeralPublicKey());
+        final byte[] encryptedData = Base64.getDecoder().decode(request.getEncryptedData());
+        final byte[] mac = Base64.getDecoder().decode(request.getMac());
+        final byte[] requestNonce = request.getNonce() != null ? Base64.getDecoder().decode(request.getNonce()) : null;
+        final Long requestTimestamp = request.getTimestamp();
+
+        // Prepare new envelope key, depending on secret's configuration.
+        final EciesEnvelopeKey envelopeKey;
+        if (encryptorSecrets.getEnvelopeKey() != null) {
+            // Envelope key is precalculated
+            envelopeKey = new EciesEnvelopeKey(encryptorSecrets.getEnvelopeKey(), ephemeralPublicKey);
+        } else {
+            // Derive shared secret from private key and ephemeral public key.
+            envelopeKey = EciesEnvelopeKey.fromPrivateKey(
+                    encryptorSecrets.getServerPrivateKey(),
+                    ephemeralPublicKey,
+                    encryptorId.getEciesSharedInfo1(encryptorParameters.getProtocolVersion())
+            );
+        }
+        // Prepare sharedInfo2 for all available information.
+        final byte[] sharedInfo2 = EciesUtils.deriveSharedInfo2(
+                encryptorParameters.getProtocolVersion(),
+                sharedInfo2Base,
+                ephemeralPublicKey,
+                requestNonce,
+                requestTimestamp,
+                associatedData
+        );
+        // Once we have SharedInfo2 prepared, we can construct a decryptor.
+        final EciesDecryptor eciesDecryptor = new EciesDecryptor(envelopeKey, sharedInfo2);
+        // Prepare EciesPayload and try to decrypt data.
+        final byte[] plaintext = eciesDecryptor.decrypt(new EciesPayload(
```

Comment:
This code is hard to read, can we please separate this code into multiple lines and introduce 2 variables for new objects?